### DISCUSSION TITLE
Soma question about use checkstyle

### DISCUSSION THREAD

**Discussion by springcoco** [checkstyle/checkstyle#12815] [Q&A]
Hello community members, in my current company we are using some self-developed frameworks internally. Due to the many versions of these frameworks, each department is using a slightly different version with some changes in functionality. Thus, we often encounter problems during upgrades and migrations. Currently, I want to use Checkstyle to implement a tool for feature scanning, in order to scan the features of the current version of the framework used by our users. However, I lack experience in code static analysis and would be grateful for any ideas or suggestions you have to offer.

**Comment by nrmancuso** [comment]
@springcoco it is not clear how we can help, please provide concrete examples of what you hope to achieve

**Reply by springcoco** [reply]
Thank you for your reply. Let me give you an example. Suppose there is a class that provides HTTP connection functionality. Version 1.0 of the framework only supports HTTP 1.0 connections, while version 2.0 supports HTTP2 connections. Now, a department is using version 1.0 of the framework and the team wants to upgrade to version 2.0. With the help of this scanning tool, they can obtain the current version's features. Specifically, after using this tool to scan, the team will know that the current version of the framework only supports HTTP 1.0 connections.

**Reply by romani** [reply]
you can start contribution to checkstyle and you will experience all static analysys tool that exists for all java related files (java,property,yml, xml, .......). Please join a team as contributor and you will master in doing it.

you do not need to apply to gsoc, but this is good start - https://github.com/checkstyle/checkstyle/wiki/To-GSoC-participant

**Reply by springcoco** [reply]
@romani 
Thank you very much for your suggestion. I think Checkstyle is a great tool and I would love to contribute to it. Currently, I am still learning. However, for my particular problem, I do not have any solutions in mind at the moment. Would you be able to provide me with some hints?

**Reply by nrmancuso** [reply]
IMO, this sounds very easy to achieve with git, and can be enhanced quite a bit by the GitHub interface:

1. Create git tag on new version in framework repo
2. In tagged SHA, author leaves GitHub comments on diff to explain changes.
3. When someone wants to update the framework version in their project, then can just visit the repo and view the tag to see changes.

**Reply by springcoco** [reply]
@nrmancuso Because I need to scan the framework version used by the current application program, Git cannot be used. What I am thinking about now is to use static scanning to achieve this function. I want to read the framework source code, build the AST, and then perform semantic analysis to determine the existing features of the current version of the framework. However, I still don't understand how to determine the functions of the current version of the framework, so I would like to ask for some advice.
